ipa-core/src/protocol/ipa_prf/validation_protocol/validation.rs

```diff
@@ -266,50 +277,44 @@ impl BatchToVerify {
     }
 
     /// This function verifies a `BatchToVerify`.
-    pub async fn verify<C, U, V>(
+    pub async fn verify<C>(
         &self,
         ctx: &C,
         sum_of_uv_right: Fp61BitPrime,
-        u_from_right_prover: U,
-        v_from_left_prover: V,
+        p_r_right_prover: Fp61BitPrime,
+        q_r_left_prover: Fp61BitPrime,
+        challenges_for_left_prover: &[Fp61BitPrime],
```

Comment:
you convinced me in #1173 that you need iterators here. Is it a TODO?

Reply:
@akoshelev everything is small except the list of all u,v values. There are 50 million u,v values, whereas there are only around 8 challenges in total and 8 "sub" proofs stored in the BatchToVerify for each left and right. So vectors are no issue here.

Reply:
ah you are changing the contract for this function, so it is no longer taking `u` and `v` but it takes challenges. 

You may consider structuring proofs so `r` values along with challenges come in one struct.